### DISCUSSION TITLE
Making a Moose function value dependent on the subdomain

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#27315] [Q&A General] [answered]
Dear Community,

For factoring material parameters I like to have some function values to depend on the subdomain. With this discussion I hope to get a hint in the right direction from you:

- Is there a way to make the function value dependent on the subdomain already build in?
- One option would be to implement my own function object (derived from [Function](framework/src/functions/Function.C)). The methods "value", "gradient", and "timeDerivative" in a (Moose) [Function](framework/src/functions/Function.C) have the arguments `Real t` and `Point & p`. But obviously no subdomainID. To avoid hard-coding my geometry to my implementation, I need to get the subdomain when calculating the function value.  Is there an efficient way to get the subdomain?
- Is the question itself feasible? What about points in space occupied by two or more subdomains (e.g. corner nodes)? The alternative would be to have each subdomain have its own material. With a lot of subdomains this would give a lot of materials. But no need for my endeavor described above. Should this alternative be preferred?

Hope someone can give me some advice. Much appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you want a dependency on the subdomain, it's better to use material properties. 
You can code any function as a material property fairly easily (use the _current... attributes indexed [_qp] for the position) and the block restriction is very natural for material properties.

If you must use Functions,
You can create several functions, and then map them to subdomains using this functor material for example:
https://mooseframework.inl.gov/source/functormaterials/PiecewiseByBlockFunctorMaterial.html
then functor materials can be converted to create regular material properties using this:
https://mooseframework.inl.gov/source/materials/MaterialFunctorConverter.html

> Is the question itself feasible? 

this is exactly why we didnt implement that dependency

Guillaume
